Culling math

Squareys opened the pull request:
Hi @mosra !

As discussed per gitter, I started working on some culling math for a project at university. This pullrequest will add cleanly tested implementations for:

 - [x] `Math::Geometry::Intersection::sphereFrustum`
 - [x] Improve `Math::Geometry::Intersection::boxFrustum`
     - [x] Deprecate and rename to `Math::Geometry::Intersection::rangeFrustum`
     - [x] Split and add to`Math::Geometry::Intersection::aabbFrustum`
 - [x] `Math::Geometry::Intersection::pointCone`
 - [x] `Math::Geometry::Intersection::pointDoubleCone`
 - [x] `Math::Geometry::Intersection::sphereCone`
 - [x] `Math::Geometry::Intersection::rangeCone`
 - ~~(Maybe `Math::Geometry::Intersection::triangleCone`, even though it's not really useful for culling)~~

Cheers, Jonathan

___

~~Plus batch intersection methods for:~~ (not part of this PR, progress backed up in [separate branch](https://github.com/Squareys/magnum/tree/culling-math-batch) )
 - [x] ~~Range - Frustum~~
 - ~~Range - Cone~~
 - [x] ~~Sphere - Frustum~~ Batching does not seem to yield improvements with array of structures
 - ~~Sphere - Cone~~

~~Should I put them into a separate file? And/or separate `Batch::` namespace?~~ Discussed: `IntersectionBatch.h/.cpp` and same namespace.

mosra commented on src/Magnum/Math/Geometry/Intersection.h:
```diff
@@ -155,6 +155,19 @@ of the frustum, since checking the corners is less efficient.
+Checks for each plane of the frustum whether the sphere is behind the plane (the
+points distance larger than the sphere's radius) using @ref Distance::pointPlaneScaled().
+*/
+template<class T> bool sphereFrustum(const Vector3<T>& center, const T radius, const Frustum<T>& frustum);
```
no need to have `const T radius` in the declaration -- only in the definition

mosra commented on src/Magnum/Math/Geometry/Intersection.h:
```diff
@@ -188,6 +201,89 @@ template<class T> bool boxFrustum(const Range3D<T>& box, const Frustum<T>& frust
+    return true;
+}
+
+/**
```
Could the functions below be moved up in front of the long function bodies to be together with the documented function declarations? Reason: people reading docs straight from header files would not need to scroll this far down.

mosra commented on src/Magnum/Math/Geometry/Intersection.h:
```diff
@@ -188,6 +201,89 @@ template<class T> bool boxFrustum(const Range3D<T>& box, const Frustum<T>& frust
+
+Returns `true` if the point is inside the cone.
+
+@see pointCone(Vector3, Vector3, Vector3, T) @todo: That ref is never going to work...
```
Does `@see @ref pointCone(const Vector3<T>&, const Vector3<T>&, T)` work? Yes, it's a bit shitty, but what can we do...

Similarly elsewhere.

mosra commented on src/Magnum/Math/Geometry/Intersection.h:
```diff
@@ -188,6 +201,89 @@ template<class T> bool boxFrustum(const Range3D<T>& box, const Frustum<T>& frust
+
+Returns `true` if the point is inside the double cone.
+
+Uses the result of precomputing @f$x = \tan^2{\theta} + 1@f$.
```
I think not having space around `@f$` might trigger some Doxygen bugs (or syntax higlighting issues in some editors). Also the space around makes it a bit more readable.

mosra commented on src/Magnum/Math/Geometry/Test/IntersectionTest.cpp:
```diff
@@ -140,6 +151,55 @@ void IntersectionTest::boxFrustum() {
+void IntersectionTest::pointCone() {
+    const Vector3 center{0.1f, 0.2f, 0.3f};
+    const Vector3 normal{0.0f, 1.0f, 0.0f};
+    const Deg<Float> angle{72.0f};
```
Could you add a typedef to just `Deg` in the typedef section above and use that instead, for consistency?

mosra commented on src/Magnum/Math/Geometry/Intersection.h:
```diff
@@ -188,6 +201,89 @@ template<class T> bool boxFrustum(const Range3D<T>& box, const Frustum<T>& frust
+@param p        The point
+@param origin   Origin of the cone
+@param normal   Normal of the cone
+@param tanAngleSqPlusOne Precomputed portion of the cone intersection equation: `Math::tan(angle/2)+1`
```
It's possible to wrap lines in `@param` blocks, doesn't need to be all on one line.

The m.css Doxygen theme makes it possible to have inline code highlighting, so you can write `@cpp Math::tan(angle/2) + 1 @ce` to have the syntax highlighted (it will also link to Math::tan() once I implement that ;)).

mosra commented on src/Magnum/Math/Geometry/Intersection.h:
```diff
@@ -188,6 +201,89 @@ template<class T> bool boxFrustum(const Range3D<T>& box, const Frustum<T>& frust
+@param normal   Normal of the cone
+@param tanAngleSqPlusOne Precomputed portion of the cone intersection equation: `Math::tan(angle/2)+1`
+
+Returns `true` if the point is inside the double cone.
```
I'm such a crazy person that I usually highlight even this: `@cpp true @ce` (it'll make it appear bold, in case you are wondering).

You are free to say "no" to this, if such extra work offends you :D

Squareys replied:
As I said often times before... I use vim 😈

mosra commented:
Oh, regarding the batch things: I think a separate `IntersectionBatch.h`/`IntersectionBatch.cpp` file would be okay, no need to create another namespace (the namespace chain is long enough already :D).

Squareys commented:
@mosra I would like to keep the benchmarks for `sphereCone` and `boxFrustum` separate to allow executing them separately. Does that make sense? I though about calling them along the lines of `IntersectionBoxFrustumBenchmark`.

mosra commented on src/Magnum/Math/Geometry/Test/IntersectionBenchmark.cpp:
```diff
@@ -0,0 +1,156 @@
+#include <tuple>
+#include <utility>
+
+namespace Magnum { namespace Math { namespace Geometry { namespace Benchmark {
```
Please keep this in the `Test` namespace, there's a lot of implicit behavior depending on that already (one is, for example, making Doxygen ignore all Test namespaces and not requiring documentation inside these, so you don't need to provide documentation blocks like below).

Squareys replied:
That is not a documentation block ;-)

mosra commented on src/Magnum/Math/Geometry/Test/IntersectionBenchmark.cpp:
```diff
@@ -0,0 +1,156 @@
+IntersectionBenchmark::IntersectionBenchmark() {
+    addBenchmarks({&IntersectionBenchmark::sphereConeNaive,
+                   &IntersectionBenchmark::sphereCone,
+                   &IntersectionBenchmark::sphereConeView}, 1000);
```
Wouldn't `100` be enough? I.e., don't make this run more than ~5 seconds on the CI and ~1 second on your machine. It's always possible to increase the iteration count using [--repeat-every](http://doc.magnum.graphics/corrade/classCorrade_1_1TestSuite_1_1Tester.html#TestSuite-Tester-command-line), if you want less variance in measured times.

Also, I think decreasing this number but increasing the loop count inside `CORRADE_BENCHMARK(n)` would help with the variance too -- so, for example 25 here and 50 there. You have to experiment, though :)

mosra commented on src/Magnum/Math/Geometry/Test/IntersectionBenchmark.cpp:
```diff
@@ -0,0 +1,156 @@
+    volatile bool b = false;
+    CORRADE_BENCHMARK(10) {
+        for(auto& cone : _cones) {
+            for(auto& sphere : _spheres) {
```
code golf (making lines shorter) -- maybe like this?

```cpp
CORRADE_BENCHMARK(10) for(auto& cone: _cones) for(auto& sphere: _spheres) {
    b = b ^ ...
}
```

Also, if you can, coding style -- `for(a: b)` instead of `for(a : b)`. There's never a space before `:`. Similarly in other places.

mosra commented on src/Magnum/Math/Geometry/Test/IntersectionBenchmark.cpp:
```diff
@@ -0,0 +1,156 @@
+
+#include <random>
+#include <tuple>
+#include <utility>
```
Coding style: ordering includes: system headers first, then dependency headers (with `<>`), then project-local headers (`""`). Thank you :smiley:

mosra commented on src/Magnum/Math/Geometry/Intersection.h:
```diff
@@ -147,14 +149,122 @@ template<class T> bool pointFrustum(const Vector3<T>& point, const Frustum<T>& f
+ * @param angle Cone opening angle (0 < angle < pi).
+ *
+ * Returns `true` if the sphere intersects with the cone.
+ */
```
No need to bother with the ` * ` at line starts if the comment block is not indented in any way. Also, `@cpp true @ce` (sorry :D)

mosra commented on src/Magnum/Math/Geometry/Test/IntersectionBoxFrustumBenchmark.cpp:
```diff
@@ -0,0 +1,121 @@
+        {0.0f, 1.0f, 0.0f, 0.0f},
+        {0.0f, -1.0f, 0.0f, 10.0f},
+        {0.0f, 0.0f, 1.0f, 0.0f},
+        {0.0f, 0.0f, -1.0f, 10.0f}};
```
Hmm. Couldn't this be written in a shorter, more descriptive way using `Matrix4::orthographicProjection({2.0f, 2.0f}, 0.0f, 10.0f)` or something?

Squareys replied:
Well, `Frustum::fromMatrix(Matrix4::orthographicProjection({2.0f, 2.0f}, 0.0f, 10.0f))` maybe.

mosra commented:
> I would like to keep the benchmarks for sphereCone and boxFrustum separate to allow executing them separately

No need to do that, the test suite is capable of handling this on its own. See the [command-line / environment options](http://doc.magnum.graphics/corrade/classCorrade_1_1TestSuite_1_1Tester.html#TestSuite-Tester-command-line), particularly the `--skip-tests` / `--skip-benchmarks` options.

mosra commented on src/Magnum/Math/Geometry/IntersectionBatch.h:
```diff
@@ -0,0 +1,100 @@
+ * @brief Namespace @ref Magnum::Math::Geometry::Intersection
+ */
+
+#include <bitset>
```
Leftover? :)

Squareys replied:
Yeah, leftover. Though about using bitset for the result, but it turns out to be compile-time defined size (or have to use dynamic_bitset)

mosra commented on src/Magnum/Math/Geometry/IntersectionBatch.h:
```diff
@@ -0,0 +1,100 @@
+
+#include <bitset>
+
+#include "Corrade/Containers/ArrayView.h"
```
Use `<>` for dependency includes, please

mosra commented on src/Magnum/Math/Geometry/Intersection.h:
```diff
@@ -167,25 +323,246 @@ template<class T> bool pointFrustum(const Vector3<T>& point, const Frustum<T>& f
+
+template<class T> bool sphereCone(
+        const Vector3<T>& sCenter, const T radius,
+        const Vector3<T>& origin, const Vector3<T>& normal, const T sinAngle, const T tanAngleSquaredPlusOne) {
```
Not sure about the indentation here. Maybe everything on one line? Or indent by 4 spaces and put the `{` on a new line?

Similarly above.

Squareys replied:
> Or indent by 4 spaces and put the { on a new line?

I like that option 👍

mosra commented on src/Magnum/Math/Geometry/Intersection.h:
```diff
@@ -147,14 +149,168 @@ template<class T> bool pointFrustum(const Vector3<T>& point, const Frustum<T>& f
+*/
+template<class T> bool aabbCone(const Vector3<T>& center, const Vector3<T>& extents, const Vector3<T>& origin, const Vector3<T>& normal, const Rad<T> angle);
+
+template<class T> bool aabbCone(const Vector3<T>& center, const Vector3<T>& extents, const Vector3<T>& origin, const Vector3<T>& normal, const T tanAngleSquaredPlusOne);
```
Since the documentation is a very important part of this, it would be great if you could generate it and verify it says what it should say.

For the overloads, a nice thing could be to say how all the extra parameters are precomputed and also referencing them from the convenience versions.

Squareys commented on src/Magnum/Math/Geometry/Intersection.h:
```diff
@@ -167,25 +323,246 @@ template<class T> bool pointFrustum(const Vector3<T>& point, const Frustum<T>& f
+
+
+template<class T> bool pointCone(const Vector3<T>& p, const Vector3<T>& origin, const Vector3<T>& normal, const Deg<T> angle) {
+    const T x = T(1)+Math::pow<T>(Math::tan<T>(angle/T(2)), T(2));
```
I am unsure why I added those (I usually wouldn't), will check!

Squareys commented on src/Magnum/Math/Geometry/Intersection.h:
```diff
@@ -167,25 +323,246 @@ template<class T> bool pointFrustum(const Vector3<T>& point, const Frustum<T>& f
+
+template<class T> bool pointDoubleCone(const Vector3<T>& p, const Vector3<T>& origin, const Vector3<T>& normal, const T tanAngleSqPlusOne) {
+    const Vector3<T> c = p - origin;
+    const T lenA = dot(c, normal);
```
Dot (c,normal) is not the squared length of a vector (I believe you may be confusing it with `c.dot()` for example).

`a` is a quantity that is not clearly defined here (I think by bachelors thesis may clear up what `a` is. It's the length of c projected onto the normal or something like that, if that makes sense.

Squareys commented:
@mosra This is *finally* ready for full review from my side! 👍 

I squashed everything together and rebased onto master. Looking forward to the review.
Cheers, Jonathan.

mosra commented on src/Magnum/Math/Geometry/Intersection.h:
```diff
@@ -140,20 +142,246 @@ points distance from the plane is negative) using @ref Distance::pointPlaneScale
+
+The points on the faces that are closest to this intersection point are the closest
+to the cone's axis and are tested for intersection with the cone using
+@ref pointCone(const Vector3<T>&, const Vector3<T>&, const Vector3<T>&, const T) "pointCone".
```
... `"pointCone()"`? I.e., so it's clear it refers to a function. 

Similarly in other cases.

mosra commented on src/Magnum/Math/Geometry/Intersection.h:
```diff
@@ -140,20 +142,246 @@ points distance from the plane is negative) using @ref Distance::pointPlaneScale
+@ref pointCone(const Vector3<T>&, const Vector3<T>&, const Vector3<T>&, const T) "pointCone".
+
+As soon as an intersecting point is found, the function returns @cpp true @ce.
+If all points lie outside of the cone, it will return @cpp false @ce.
```
At first I thought this would give a "false negative" in case the cone is so narrow that it fits inside the AABB without touching the edges. But that's handled by the above mentioned plane intersection, right? So maybe update the sentence to make it clear this case is handled correctly.

mosra commented on src/Magnum/Math/Geometry/Intersection.h:
```diff
@@ -166,28 +394,206 @@ template<class T> bool pointFrustum(const Vector3<T>& point, const Frustum<T>& f
 
-        for(UnsignedByte c = 0; c != 8; ++c) {
-            const Vector3<T> corner = Math::lerp(box.min(), box.max(), Math::BoolVector<3>{c});
+    for(const Vector4<T>& plane: frustum.planes()) {
```
Why is this not just calling into the other `rangeFrustum()` overload? The code is the same.

The `box` param should be `range` to be consistent with the declaration above.

Squareys replied:
Because it's comapring with `-T(2)*plane.w()` instead of `-plane.w()`, to avoid the vector/scalar division when converting range to center & extent representation.

mosra replied:
(As discussed offline: okay, leave it as is.)

mosra commented on src/Magnum/Math/Geometry/Intersection.h:
```diff
@@ -166,28 +394,206 @@ template<class T> bool pointFrustum(const Vector3<T>& point, const Frustum<T>& f
+
+template<class T> bool sphereFrustum(
+    const Vector3<T>& center, const T radius, const Frustum<T>& frustum)
+{
```
The declarations aren't too long to be wrapped like this, I think. Similarly in most cases below.

Squareys replied:
From the Corrade [coding style](http://doc.magnum.graphics/corrade/corrade-coding-style.html) I read:

> vertical and horizontal compression, fitting more code on a screen while still making it possible to have more files open side by side by wrapping on the 80th column

I am pretty sure I only wrapped those which exceeded that quite a bit. You may need to specify "aren't too long" a bit more.

Squareys replied:
This particular line is 105 colums long if not wrapped

Squareys replied:
I did, however, miss quite a few (like the function above) it seems :D

mosra replied:
Okay, yeah, the coding style is a bit flaky here. In particular, I don't *demand* strict 78 column wrapping like for example the Python PEP-8 tools, I require that only for documentation blocks and bigger blocks of wide code, where the editor wrapping them would make the block significantly harder to read.

Function signatures being a bit longer are okay, because that's just a single line among usually short lines, unless they are so long that an editor wraps them on three and more lines, making editing and reviewing the changes harder.

All function parameter lists that are here wrapped on one or one and half line are still in the okay range to have the full signature in a single line, I think.

Squareys replied:
As in

~~~
someFunc(
    This is, One line)
{
~~~

and

~~~
SomeOtherFunc(
    That is, Clearly way, Too long,
    Not to, Be broken) //two lines? One and a half?
{
~~~

mosra commented on src/Magnum/Math/Geometry/Intersection.h:
```diff
@@ -166,28 +394,206 @@ template<class T> bool pointFrustum(const Vector3<T>& point, const Frustum<T>& f
+    const Vector3<T>& sCenter, const T radius,
+    const Vector3<T>& origin, const Vector3<T>& normal, const Rad<T> angle)
+{
+
```
Superfluous empty line.

Similarly below.

mosra commented on src/Magnum/Math/Geometry/Test/IntersectionTest.cpp:
```diff
@@ -124,20 +144,192 @@ void IntersectionTest::pointFrustum() {
+    CORRADE_VERIFY(Intersection::pointCone(center + normal, center, normal, angle));
+    /* Point outside */
+    CORRADE_VERIFY(!Intersection::pointCone(Vector3{}, center, normal, angle));
+    //TODO: Why does this fail? CORRADE_VERIFY(!Intersection::pointCone(center + 5.0f*surface + 0.1f*tangent, center, normal, angle));
```
Is this TODO still relevant?

Squareys replied:
Yes, I should have a look at that, (sorry)

Squareys replied:
![image](https://user-images.githubusercontent.com/1889284/38379673-b61d513e-3901-11e8-8ac5-c536ac6c56a3.png)
Well, not outside after all. Should be, though, wanted it to "just" not touch the surface.

mosra commented on src/Magnum/Math/Geometry/Intersection.h:
```diff
@@ -140,20 +142,246 @@ points distance from the plane is negative) using @ref Distance::pointPlaneScale
+
+@see @ref sphereConeView(const Vector3<T>&, T, const Matrix4<T>&, T, T, T)
+*/
+template<class T> bool sphereConeView(const Vector3<T>& sphereCenter, T radius, const Matrix4<T>& coneView, Rad<T> angle);
```
Does this work as expected even with the pathological matrices correctly? Such as a skew matrix or non-uniform scaling?

If not, I'd maybe use a some more restricted representation ... or at least mention it in the docs.

Squareys replied:
It will not, it expect it to be a rigid transform with no or uniform scaling.
Restricted representation as in rotation matrix + translation? Or Dual quaternion? Or Assert on [isRigidTransformation()](http://doc.magnum.graphics/magnum/classMagnum_1_1Math_1_1Matrix3.html#a267cce4346f18620cd93df07bf858234)?

mosra replied:
Hmm. What would be the best in terms of usability? 

- Allowing Matrix4 with no check inside would lead to surprising results very often, considering scaling is usually part of a transformation. 3x3 rotation matrix suffers from the same problem.
- Assert on `isRigidTransformation()` might be overly expensive
- Requiring dual quaternion might feel too alien to some people

Hmm...

Squareys replied:
Considering this is used (if at all) with either camera transformation matrices or light transformation matrices, I would expect the isRigidTransformation to be the most sensible as all other would require a conversion.

mosra commented:
Oh, looked at the benchmarks -- the difference between `sphereFrustum()` and `sphereCone()` is quite significant, maybe it would be worth to mention `sphereCone()` in the docs of `sphereFrustum()` :)

Squareys commented on src/Magnum/Math/Geometry/Intersection.h:
```diff
@@ -140,20 +142,246 @@ points distance from the plane is negative) using @ref Distance::pointPlaneScale
+for plane normal @f$ \boldsymbol n @f$ and determinant @f$ w @f$.
+
+@see @ref aabbFrustum()
+*/
```
Well, but it doesn't call aabbFrustum.

mosra replied:
But it *could* -- see [my other comment](https://github.com/mosra/magnum/pull/226/files/3055fd259ff9160861cdc541af527ef303f3d656#r177785355) .. or am I missing something?

Squareys commented on src/Magnum/Math/Geometry/Intersection.h:
```diff
@@ -140,20 +142,246 @@ points distance from the plane is negative) using @ref Distance::pointPlaneScale
+
+@see @ref pointCone(const Vector3&, const Vector3&, const Vector3&, T)
+*/
+template<class T> bool pointCone(const Vector3<T>& p, const Vector3<T>& origin, const Vector3<T>& normal, Rad<T> angle);
```
`coneNormal` is the normal of the cone, from the apex towards the open portion:
~~~
  /
 /
<-----> (n)
 \
  \
~~~

Squareys commented:
Regarding the benchmark results: I had a closer look at that and would not infer too much on that yet. The frustum and the cone are not very similar at the moment (I will fix that), so this may be the sphereFrustum test running into some worst- or worser-case maybe.

That reminds me, though: For camera frustums and view culling it is common to skip the test against the near plane. Just something to keep in mind and maybe support somehow in the future (e.g. by Frustum only having 5 planes.)

Squareys commented:
I cleaned up the benchmarks a bit. While the `sphereCone` _seems_ indeed quite a bit faster than the `sphereFrustum` on AppVeyor, which builds **Debug** config, when I build it in **Release** config locally, `sphereFrustum` is actually a little bit faster even.

I finished incorporating the review comments apart from the three that GitHub correctly didn't collapse.

mosra commented on src/Magnum/Math/Geometry/Test/IntersectionTest.cpp:
```diff
@@ -124,20 +144,192 @@ void IntersectionTest::pointFrustum() {
+    CORRADE_VERIFY(Intersection::rangeFrustum(Range3D{{-0.1f, 0.4f, 4.9f}, {0.1f, 0.6f, 5.1f}}, frustum));
+    CORRADE_VERIFY(Intersection::rangeFrustum(Range3D{{4.9f, 0.4f, 4.9f}, {5.1f, 0.6f, 5.1f}}, frustum));
+    CORRADE_VERIFY(Intersection::rangeFrustum(Range3D{{2.4f, 0.4f, -0.1f}, {2.6f, 0.6f, 0.1f}}, frustum));
+    CORRADE_VERIFY(Intersection::rangeFrustum(Range3D{{2.4f, 0.4f, 9.9f}, {2.6f, 0.6f, 10.1f}}, frustum));
```
What about using Range::fromSize() in the complex cases here? Would make it quite easier to maintain.

Squareys replied:
Yeah, that would condense this a bit, I hope.

Squareys commented:
@mosra Done, thanks for all the feedback! This is once again ready for review :)

mosra commented:
Attempted to merge this, but:

- Can you rebase on current master to let all the CIs run through? I'm not sure the `CORRADE_EXPECT_FAIL()` would cover so many test cases on Emscripten.
- You can add `Vector3d` and `Radd` typedefs to the test instead of using the verbose `Math::Vector3<Double>` there (and also can omit the `<Double>` for the function call)
- There is some trailing whitespace in the test
- The `sphereConeView()` function is called 640000 times from the benchmark, but not even once from the test. Also, can you add the assert for `isRigidTransformation()` there as discussed above (and a test case with some pathological matrix)? (TODO for me: don't allow benchmarks to contribute to code coverage)
- Some of my comments above are not resolved yet ([1](https://github.com/mosra/magnum/pull/226#discussion_r177779392), [2](https://github.com/mosra/magnum/pull/226#discussion_r177779606))

Thanks!

mosra commented:
~~One~~ ~~two~~ ~~three~~ ~~four~~ five more things, sorry that I'm not commenting in the code directly, the GitHub UI is awfully slow at this point :/

1. The `sphereConeView()` function now has the `cosAngle` parameter unused (code seems to changed compared to last time?):

  - Is that intentional? Shouldn't the function take one parameter less now?
  - In any case I don't think it's needed to force the user to specify all three of sin, cos and tan (one can be calculated from the other two)

2. The `pointCone()` test has `centerDouble`, `normalDouble` and `angleDouble` variables, none of which is used. Is there something missing? Those variables seem to be used in other tests, though.

3. some functions mention that 0 < cone opening angle < pi, but some not. Should I copy that everywhere for consistency? Is the same for "Apex angle of the cone" in the double cone case? Shouldn't that be also "Cone opening angle"? Or shouldn't be all "cone opening angles" rather "apex angles"? 
And an uppercase Greek letter Theta for this? (That's how Wikipedia has it, at least. You're using lowercase Theta for the half-angle.)

4. Looking at the code, basically all functions taking the apex angle (or whatever the name is) are cutting it in half. What about changing them to already take half angle and save one multiplication there? This will be easy to change now, but *hard* to change later.

5. Both `rangeCone()` functions calculate the same thing and then call `aabbCone()`. Shouldn't the first one (taking the angle) call the second one (taking tanAngleSqPlusOne) and then the second one call `aabbCone()`? Or did I miss something?

Since I'm quite far into the merge (had to rename function parameters for consistency with docs and fix a bunch of errors in the math formulas), I can do the above changes myself (otherwise I would get lots of nasty conflicts later), just tell me how ;)

(Note to myself, in case I forget again: I have an in-progress merge in a private `intersection-merge` branch)

Squareys commented:
1. Yeah `cosAngle` can be removed
2. A "point on cone" test would still be missing, I checked that locally and it's this test case:
~~~cpp
    // after auto sNormal = Matrix4::rotation(90.0_degf, axis).transformVector(surface);

    /* Same for Double precision */
    auto axisDouble = Math::cross(Vector3d::yAxis(), normalDouble).normalized();
    auto surfaceDouble = Matrix4d::rotation(0.5*angleDouble, axisDouble).transformVector(normalDouble);

    // ...

    /* Point touching cone */
    CORRADE_VERIFY(Intersection::pointCone(centerDouble, centerDouble + surfaceDouble, normalDouble, angleDouble));

~~~

*But* this test fails similar to the others, e.g. the sphere touching cone test, possibly because of precision. Also, the whether emscripten fails or not, I don't know.

3. Yes, it's the same opening angle everywhere and only "acute cones" (hence, 0 < angle < pi). Whether you call it opening angle or apex angle is not too important, I think. On wikipedia it's "apex angle", so maybe you want to use that consistently. For the angle notation just follow the convention you normally use for angles.

4. That's only for the non-precomputed slow function anyway. If you really want to save the multiplication, you can precompute it yourself. This one should be convenient, and I believe that specifying the entire angle instead of the half angle is more intuitive. I'm fine with documenting that it should be the half angle, too, though, if you prefer that.

5. Yes, indeed!

Thank you for all the effort! And sorry that this seems to be producing a lot of work :/

mosra commented:
1. done :)
2. this seems to be failing similarly to the other. Also it looks like it does not fail on the CI Emscripten but it does on my machine. Ugh.
3. changed to "apex angle" everywhere and using uppercase Theta to be consistent with Wikipedia.
4. good point :+1: does not make sense to optimize anything here

Once I resolve no.2, Is it okay if I push the changes to your branch so you can check I did not do anything terribly wrong to the code?

mosra commented:
Pushed :) If you can take a quick look, just to have more than one pair of eyes seeing the changes before putting them to master forever, that would be great.

Squareys commented:
Looks good to me, thanks! 👍

mosra commented:
Aaaand [you're shipped](http://doc.magnum.graphics/magnum/namespaceMagnum_1_1Math_1_1Intersection.html#a796a25abb39fdaecab9933aff6c70836)! :tada: 

Thanks a lot for this huge piece of work.

Squareys commented:
Hooray, awesome, thanks! 🎉
